pkg/services/ngalert/accesscontrol/receivers.go

```diff
@@ -0,0 +1,174 @@
+	provisioningExtraReadRedactedPermissions = ac.EvalAny(
+		ac.EvalPermission(ac.ActionAlertingProvisioningRead),              // Global provisioning action for all AM config. Org scope.
+		ac.EvalPermission(ac.ActionAlertingNotificationsProvisioningRead), // Global provisioning action for receivers. Org scope.
+		provisioningExtraReadDecryptedPermissions,
+	)
+
+	// Extra permissions that give read-only access to all decrypted receivers when called from provisioning api.
+	provisioningExtraReadDecryptedPermissions = ac.EvalAny(
+		ac.EvalPermission(ac.ActionAlertingProvisioningReadSecrets), // Global provisioning action for all AM config + secrets. Org scope.
+	)
+)
+
+type ReceiverAccess[T models.Identified] struct {
+	read          actionAccess[T]
+	readDecrypted actionAccess[T]
+}
+
+// NewReceiverAccess creates a new ReceiverAccess service. If includeProvisioningActions is true, the service will include
+// permissions specific to the provisioning API.
+func NewReceiverAccess[T models.Identified](a ac.AccessControl, includeProvisioningActions bool) *ReceiverAccess[T] {
+	rcvAccess := &ReceiverAccess[T]{
+		read: actionAccess[T]{
+			genericService: genericService{
+				ac: a,
+			},
+			resource:      "receiver",
+			action:        "read",
+			authorizeSome: readRedactedReceiversPreConditionsEval,
+			authorizeOne: func(receiver T) ac.Evaluator {
+				return readRedactedReceiverEval(receiver.GetUID())
+			},
+			authorizeAll: readRedactedAllReceiversEval,
+		},
+		readDecrypted: actionAccess[T]{
+			genericService: genericService{
+				ac: a,
+			},
+			resource:      "decrypted receiver",
+			action:        "read",
+			authorizeSome: readDecryptedReceiversPreConditionsEval,
+			authorizeOne: func(receiver T) ac.Evaluator {
+				return readDecryptedReceiverEval(receiver.GetUID())
+			},
+			authorizeAll: readDecryptedAllReceiversEval,
+		},
+	}
+
+	// If this service is meant for the provisioning API, we include the provisioning actions as possible permissions.
+	if includeProvisioningActions {
+		rcvAccess.read.authorizeSome = ac.EvalAny(provisioningExtraReadRedactedPermissions, rcvAccess.read.authorizeSome)
+		rcvAccess.readDecrypted.authorizeSome = ac.EvalAny(provisioningExtraReadDecryptedPermissions, rcvAccess.readDecrypted.authorizeSome)
+
+		rcvAccess.read.authorizeOne = func(receiver T) ac.Evaluator {
+			return ac.EvalAny(provisioningExtraReadRedactedPermissions, rcvAccess.read.authorizeOne(receiver))
+		}
+		rcvAccess.readDecrypted.authorizeOne = func(receiver T) ac.Evaluator {
+			return ac.EvalAny(provisioningExtraReadDecryptedPermissions, rcvAccess.readDecrypted.authorizeOne(receiver))
+		}
+
+		rcvAccess.read.authorizeAll = ac.EvalAny(provisioningExtraReadRedactedPermissions, rcvAccess.read.authorizeAll)
+		rcvAccess.readDecrypted.authorizeAll = ac.EvalAny(provisioningExtraReadDecryptedPermissions, rcvAccess.readDecrypted.authorizeAll)
+	}
```

Comment:
I agree that will be better. It's a bit awkward to do in this PR since there is some legacy overlap where the receiver api uses provisioning actions. I remove them in https://github.com/grafana/grafana/pull/90861 so I'll move to the composition approach after that.